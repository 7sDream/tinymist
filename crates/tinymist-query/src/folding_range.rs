use crate::{get_lexical_hierarchy, prelude::*, LexicalScopeGranularity};

#[derive(Debug, Clone)]
pub struct FoldingRangeRequest {
    pub path: PathBuf,
}

impl FoldingRangeRequest {
    pub fn request(
        self,
        world: &TypstSystemWorld,
        position_encoding: PositionEncoding,
    ) -> Option<Vec<FoldingRange>> {
        let source = get_suitable_source_in_workspace(world, &self.path).ok()?;

        let symbols = get_lexical_hierarchy(source, LexicalScopeGranularity::Block)?;

        let _ = symbols;
        let _ = position_encoding;

        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::*;

    #[test]
    fn test_folding_range_request() {
        run_with_source("let a = 1;", |world, path| {
            let request = FoldingRangeRequest { path };
            let result = request.request(world, PositionEncoding::Utf16);
            assert_eq!(result, None);
        });
    }
}
