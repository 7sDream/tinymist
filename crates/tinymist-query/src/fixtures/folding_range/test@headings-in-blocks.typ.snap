---
source: crates/tinymist-query/src/folding_range.rs
expression: "JsonRepr::new_pure(result.unwrap())"
input_file: crates/tinymist-query/src/fixtures/folding_range/headings-in-blocks.typ
---
[
 {
  "collapsedText": "Heading 2",
  "endCharacter": 16,
  "endLine": 8,
  "startCharacter": 7,
  "startLine": 5
 },
 {
  "collapsedText": "",
  "endCharacter": 3,
  "endLine": 7,
  "startCharacter": 11,
  "startLine": 3
 },
 {
  "collapsedText": "Heading 1",
  "endCharacter": 14,
  "endLine": 8,
  "startCharacter": 5,
  "startLine": 1
 },
 {
  "collapsedText": "",
  "endCharacter": 1,
  "endLine": 8,
  "startCharacter": 9,
  "startLine": 0
 }
]
